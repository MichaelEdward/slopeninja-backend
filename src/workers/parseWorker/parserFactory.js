import cheerio from 'cheerio';

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html));

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser) => async (data = "{}") => {
  let json = JSON.parse(data);

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
