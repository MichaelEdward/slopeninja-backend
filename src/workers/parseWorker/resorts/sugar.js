import cheerio from 'cheerio';

import {
  degreeOrNull,
  inchOrNull,
} from '../util';

const initialWeather = {
  status: null,
  weatherIcon: null,
  temprature: null,
  baseCondition: null,
  newSnow: null,
  snowDepthBase: null,
  snowDepthSummit: null,

};

const parseSugar = async ($) => {
  const temprature = $('#container_314_outer .conditions_col .table_text_01.c4').first().text().trim();
  //24 Hours
  const newSnow24Hr = $('#container_313_outer .conditions_col .table_text_01.c4').slice(3,4).text().trim();
  //Base
  const snowDepthBase = $('#container_314_outer .conditions_col.conditions_col_break1 .table_text_01.c4').text().trim();

  const snowDepthSummit = $('#container_313_outer .conditions_col.conditions_col_break1 .table_text_01.c4').text().trim();

  return {
    ...initialWeather,
    temprature: degreeOrNull(temprature),
    newSnow: inchOrNull(newSnow24Hr),
    snowDepthBase: inchOrNull(snowDepthBase),
    snowDepthSummit: inchOrNull(snowDepthSummit),
  };
}

export const fetchSugar = async (html) => {
  const $ = cheerio.load(html)

  const weather = await parseSugar($);

  return {
    weather,
  };
}
