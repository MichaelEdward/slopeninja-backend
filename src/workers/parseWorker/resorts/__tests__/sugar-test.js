import fs from 'fs';
import { parseSugarWeather, parseSugarLifts, parseSugarTrails } from '../sugar';
import { createHtmlParser } from '../../parserFactory';

test('fetches Sugar weather data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/sugar-weather.html`);

  const resortData = await createHtmlParser('weather', parseSugarWeather)(htmlText);
  expect(resortData).toEqual({
    weather: {
      status: 'closed',
      weatherIcon: 'snow',
      temprature: 32,
      baseCondition: null,
      newSnow: 7,
      snowDepthBase: 87,
      snowDepthSummit: 195,
    }
  });
})

test('fetches all null for nonexisting weather values', async () => {
  const resortData = await createHtmlParser('weather', parseSugarWeather)('<html></html>');
  expect(resortData).toEqual({
    weather: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Sugar lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/sugar-weather.html`);
  const resortData = await createHtmlParser('lifts', parseSugarLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseSugarLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Sugar trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/sierra-lifts.html`);
  const resortData = await createHtmlParser('trails', parseSugarTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseSugarTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});
