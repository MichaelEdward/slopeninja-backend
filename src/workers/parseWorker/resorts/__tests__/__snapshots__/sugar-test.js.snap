// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Sugar lift list correctly 1`] = `
Object {
  "lifts": Array [
    Object {
      "category": "Village Portal",
      "name": "Lincoln Express",
      "status": "closed",
    },
    Object {
      "category": "Village Portal",
      "name": "Disney Express",
      "status": "closed",
    },
    Object {
      "category": "Village Portal",
      "name": "Christmas Tree Express",
      "status": "closed",
    },
    Object {
      "category": "Village Portal",
      "name": "Nob Hill",
      "status": "closed",
    },
    Object {
      "category": "Village Portal",
      "name": "Crow's Peak",
      "status": "closed",
    },
    Object {
      "category": "Judah Portal",
      "name": "White Pine",
      "status": "closed",
    },
    Object {
      "category": "Judah Portal",
      "name": "Judah Express",
      "status": "closed",
    },
    Object {
      "category": "Judah Portal",
      "name": "Summit Chair",
      "status": "closed",
    },
    Object {
      "category": "Judah Portal",
      "name": "Jerome Hill Express",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Sugar trail list correctly 1`] = `
Object {
  "trails": Array [
    Object {
      "category": "Lincoln Express",
      "level": "expert",
      "name": "Fuller's Folly",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "expert",
      "name": "Hidden Gully",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "expert",
      "name": "Sisters",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "expert",
      "name": "The '58",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "expert",
      "name": "The Palisades",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Vanderbilt",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "East Face",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Sherpa Creek",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Steilhang",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Steilhang Gully",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Silver Belt Run",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Carl's Nose",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "advanced",
      "name": "Hari-Kari Gully",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Hellman's Chute",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Henderson's Bowl",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Bill Klein's Schuss",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "California Street",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Lake View",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Chick's",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Crowley's",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Crowley's Fingers",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Rahlves' Run (upper)",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Short Cut",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Station B",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Rahlves' Run (lower)",
      "status": "closed",
    },
    Object {
      "category": "Lincoln Express",
      "level": "intermediate",
      "name": "Union Pacific",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Disney Nose",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Donald Duck",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Eagle",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "East Face",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Hour Glass",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Avalanche",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Bacon's Gully",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Market Street",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Sugar Bowl",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "advanced",
      "name": "Nancy's Couloir",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Overland",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Pony Express",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Montgomery",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Upper MacTavish",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Disney Traverse",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Disney Meadow",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Lonesome Pine",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Lower MacTavish",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "LP Fingers",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "Mac R/O",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "intermediate",
      "name": "John John's Way",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "beginner",
      "name": "Meadow Run",
      "status": "closed",
    },
    Object {
      "category": "Disney Express",
      "level": "beginner",
      "name": "Disney Return",
      "status": "closed",
    },
    Object {
      "category": "Nob Hill",
      "level": "beginner",
      "name": "Nob Hill Run",
      "status": "closed",
    },
    Object {
      "category": "Nob Hill",
      "level": "beginner",
      "name": "Union Street",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "advanced",
      "name": "Blitzen",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Upper Ridge",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Van Ruiten's Run",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Slot",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Chase Ridge",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Mistletoe",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Ridge Run",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "intermediate",
      "name": "Silver Belt Run-out",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "beginner",
      "name": "Sleigh Ride",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "beginner",
      "name": "Christmas Tree Lane",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "beginner",
      "name": "Harriet's Hollow",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "beginner",
      "name": "Judah Bound",
      "status": "closed",
    },
    Object {
      "category": "Christmas Tree Express",
      "level": "beginner",
      "name": "Catwalk",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "expert",
      "name": "Crow's Face",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "expert",
      "name": "Crow's Nest Glades",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "expert",
      "name": "The Other One",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "advanced",
      "name": "Strawberry Fields",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "advanced",
      "name": "Crow's Traverse",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "advanced",
      "name": "Rob's Run",
      "status": "closed",
    },
    Object {
      "category": "Crow's Peak",
      "level": "intermediate",
      "name": "Lower Overland",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "expert",
      "name": "Tunnel 41 Upper",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "advanced",
      "name": "Emigrant Glades",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "advanced",
      "name": "Juniper",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "advanced",
      "name": "Pacific Crest Glades",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Pacific Crest Trail",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Sunset (lower)",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Sunset (upper)",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Golden Gate",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Buena Vista",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Caboose",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Coldstream",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "River Run Gully",
      "status": "closed",
    },
    Object {
      "category": "Judah Express",
      "level": "intermediate",
      "name": "Tunnel 41 Lower",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "advanced",
      "name": "Donner's Way",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "advanced",
      "name": "Graydon Glades",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "advanced",
      "name": "Steamer's",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Trailblazer",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Emigrant Gap",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Broken Axle",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Caleb's Way",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Central Pacific",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "intermediate",
      "name": "Sidewinder",
      "status": "closed",
    },
    Object {
      "category": "Jerome Hill Express",
      "level": "beginner",
      "name": "Pioneer Trail",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "expert",
      "name": "Tunnel",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "expert",
      "name": "Gold Rush",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "expert",
      "name": "Railroad Run",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "expert",
      "name": "Century Club",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "expert",
      "name": "Judah Bowl",
      "status": "closed",
    },
    Object {
      "category": "Summit Chair",
      "level": "advanced",
      "name": "Roller Pass",
      "status": "closed",
    },
    Object {
      "category": "White Pine",
      "level": "beginner",
      "name": "Black Bear",
      "status": "closed",
    },
    Object {
      "category": "White Pine",
      "level": "beginner",
      "name": "Learning Zone",
      "status": "closed",
    },
    Object {
      "category": "White Pine",
      "level": "beginner",
      "name": "Marmot",
      "status": "closed",
    },
    Object {
      "category": "White Pine",
      "level": "beginner",
      "name": "Parking Return",
      "status": "closed",
    },
    Object {
      "category": "White Pine",
      "level": "beginner",
      "name": "Pine Marten",
      "status": "closed",
    },
  ],
}
`;
